CMakeLists.txt

```diff
@@ -1818,6 +1821,80 @@ if(USE_DISCORD AND NOT IOS AND NOT LIBRETRO)
+if(USE_MINIUPNPC)
+	set (MINIUPNPC_VERSION 2.1) # used by miniupnpcstrings.h.cmake
+	set (MINIUPNPC_API_VERSION 17)
+	option(UPNPC_BUILD_STATIC "Build static library" TRUE)
+	option(NO_GETADDRINFO "Define NO_GETADDRINFO" FALSE)
+	mark_as_advanced(NO_GETADDRINFO)
+	if (NO_GETADDRINFO)
+	  add_definitions(-DNO_GETADDRINFO)
+	endif()
+	
+	if (NOT WIN32)
+	  add_definitions (-DMINIUPNPC_SET_SOCKET_TIMEOUT)
+	  add_definitions (-D_BSD_SOURCE -D_DEFAULT_SOURCE -D_POSIX_C_SOURCE=200112L)
+	else()
+	  add_definitions (-D_WIN32_WINNT=0x0501) # XP or higher for getnameinfo and friends
+	endif()
+	if (MACOSX)
+	  add_definitions (-D_DARWIN_C_SOURCE)
+	endif()
+	if(WIN32)
+		add_definitions(-DWIN32 -DMINIUPNP_EXPORTS )
+	else()
+		add_definitions(-fPIC)
+	endif()
+	
+	add_definitions(-DMINIUPNP_STATICLIB)
+	set(MINIUPNP_DIR "ext/miniupnp/miniupnpc")
+	include_directories(${CMAKE_CURRENT_BINARY_DIR})
+	configure_file(${MINIUPNP_DIR}/miniupnpcstrings.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/miniupnpcstrings.h) # by default miniupnp repo doesn't contains miniupnpcstrings.h and need to be generated
+	set(MINIUPNPC_SOURCES
+        # the needed bits of miniupnpc (no python module, no tests, no cli)
+		${MINIUPNP_DIR}/connecthostport.c
+        ${MINIUPNP_DIR}/igd_desc_parse.c
+        ${MINIUPNP_DIR}/minisoap.c
+        ${MINIUPNP_DIR}/minissdpc.c
+        ${MINIUPNP_DIR}/miniupnpc.c
+        ${MINIUPNP_DIR}/miniwget.c
+        ${MINIUPNP_DIR}/minixml.c
+        ${MINIUPNP_DIR}/minixmlvalid.c
+		${MINIUPNP_DIR}/portlistingparse.c
+		${MINIUPNP_DIR}/receivedata.c
+		#${MINIUPNP_DIR}/upnpc.c # causing an error due to already existing _main()
+        ${MINIUPNP_DIR}/upnpcommands.c
+		${MINIUPNP_DIR}/upnpdev.c
+        ${MINIUPNP_DIR}/upnperrors.c
+        ${MINIUPNP_DIR}/upnpreplyparse.c
+		${CMAKE_CURRENT_BINARY_DIR}/miniupnpcstrings.h
+	)
+	if (NOT WIN32 AND NOT CMAKE_SYSTEM_NAME STREQUAL "AmigaOS")
+	  #set(MINIUPNPC_SOURCES ${MINIUPNPC_SOURCES} minissdpc.c) # causing an error due to duplication in MINIUPNPC_SOURCES?
+	endif()
+	if (WIN32)
+	  set_source_files_properties(${MINIUPNPC_SOURCES} PROPERTIES  COMPILE_DEFINITIONS "MINIUPNP_STATICLIB;MINIUPNP_EXPORTS")
+	  find_library(WINSOCK2_LIBRARY NAMES ws2_32 WS2_32 Ws2_32)
+	  find_library(IPHLPAPI_LIBRARY NAMES iphlpapi)
+	  set(LDLIBS ${WINSOCK2_LIBRARY} ${IPHLPAPI_LIBRARY} ${LDLIBS})
+	#elseif (CMAKE_SYSTEM_NAME STREQUAL "Solaris")
+	#  find_library (SOCKET_LIBRARY NAMES socket)
+	#  find_library (NSL_LIBRARY NAMES nsl)
+	#  find_library (RESOLV_LIBRARY NAMES resolv)
+	#  set (LDLIBS ${SOCKET_LIBRARY} ${NSL_LIBRARY} ${RESOLV_LIBRARY} ${LDLIBS})
+	endif()
+	if (MSVC)
+		# Suppress noise warnings
+		target_compile_definitions(miniupnpc _CRT_SECURE_NO_WARNINGS _WINSOCK_DEPRECATED_NO_WARNINGS)
+	endif()
+	if (UPNPC_BUILD_STATIC)
+		add_library(miniupnpc STATIC ${MINIUPNPC_SOURCES})
+		target_link_libraries(${CoreLibName} miniupnpc ${LDLIBS})
```

Comment:
Does anyone know why Android LIBRETRO is the only successful build on travisci while the rest of Android compiles failed to recognize PortManager.h/cpp and getting these kind of error?
```
[arm64-v8a] SharedLibrary  : libppsspp_jni.so

./obj/local/arm64-v8a/libppsspp_core.a(sceNet.o): In function `__NetInit()':

/home/travis/build/hrydgard/ppsspp/android/jni/../../Core/HLE/sceNet.cpp:256: undefined reference to `g_PortManager'
```

Reply:
I'll look into it later today.

Reply:
This is really weird. I was able to build it just fine locally, but on CI, I don't see PortManager.cpp in the build log so it seems it just isn't building it?? Bizarre...

Reply:
As for the Windows build, in miniupnpc.vcxproj, you're gonna have to add this workaround in a few places (manually, using a text editor):

```
    <Import Project="..\Windows\fix_2017.props" />
```

See ext/glslang.vcxproj .

Reply:
Ah I just realized what's wrong with Android.

We still have an old Android.mk based build which we are using on CI. It's probably time to rip it out, but in the meantime, you need to add PortManager.cpp to android/jni/Android.mk . You'll figure out how.

Reply:
I've done both of them, but on android it still have the same error, while on Windows... i think it's getting worse LOL

Reply:
do i need to add these kind of line on Android.mk ?
```
$(call import-module,miniupnpc)
```

and also need to add miniupnpc to Locals.mk ?

Reply:
Oh, right hehe. Yes you might have to do that and also add an Android.mk for it (alternatively, just add the files directly from one of the other Android.mk that are sprinkled around).

Really we probably should just kill it, though will need to reconfigure that build bot. Will open an issue.

Reply:
One of the source file need to be generated like this on CMakeLists.txt
```
configure_file(${MINIUPNP_DIR}/miniupnpcstrings.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/miniupnpcstrings.h) # by default miniupnp repo doesn't contains miniupnpcstrings.h and need to be generated
```
Does Android.mk can also generate that file?